setuptools/tests/config/test_pyprojecttoml_dynamic_deps.py

```diff
@@ -0,0 +1,75 @@
+import pytest
+
+from setuptools.config.pyprojecttoml import apply_configuration
+from setuptools.dist import Distribution
+
+
+def test_dynamic_dependencies(tmp_path):
+    (tmp_path / "requirements.txt").write_text("six\n  # comment\n")
+    pyproject = (tmp_path / "pyproject.toml")
+    pyproject.write_text("""
+[project]
+name = "myproj"
+version = "1.0"
+dynamic = ["dependencies"]
+
+[build-system]
+requires = ["setuptools", "wheel"]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools.dynamic.dependencies]
+file = ["requirements.txt"]
+""")
+    dist = Distribution()
+    dist = apply_configuration(dist, pyproject)
+    assert dist.install_requires == ["six"]
+
+
+def test_dynamic_optional_dependencies(tmp_path):
+    (tmp_path / "requirements-docs.txt").write_text("sphinx\n  # comment\n")
+    pyproject = (tmp_path / "pyproject.toml")
+    pyproject.write_text("""
+[project]
+name = "myproj"
+version = "1.0"
+dynamic = ["optional-dependencies"]
+
+[tool.setuptools.dynamic.optional-dependencies.docs]
+file = ["requirements-docs.txt"]
+
+[build-system]
+requires = ["setuptools", "wheel"]
+build-backend = "setuptools.build_meta"
+""")
+    dist = Distribution()
+    dist = apply_configuration(dist, pyproject)
+    assert dist.extras_require == {"docs": ["sphinx"]}
+
+
+@pytest.mark.xfail(reason="PEP 621 forbids mixing static and dynamic fields")
+def test_mixed_dynamic_optional_dependencies(tmp_path):
+    # This test would pass with flying colors if `validate_project_dynamic` didn't
+    # complain about the mixed dynamic/static configuration.
+    # Perhaps PEP 621 should be loosened regarding fields containing "sub-fields"
+    # (i.e. optional dependency groups in this case)?
```

Comment:
I agree that PEP 621 is quite heavy-handed in terms of validations for `dynamic` fields,
but I don't think `setuptools` should unilaterally loosen the restrictions, that might create a non favourable reaction in the community.

All changes in the PEP behaviour can be discussed in https://discuss.python.org/c/packaging/14, but that maybe be a long and laborious process.